newton/_src/sim/builder.py

```
@@ -6315,32 +6468,64 @@ def finalize(self, device: Devicelike | None = None, requires_grad: bool = False
+                    attr_count = len(custom_attr.values) if custom_attr.values else 0
+                    if freq_key not in custom_frequency_counts:
+                        custom_frequency_counts[freq_key] = attr_count
+                        custom_frequency_attrs[freq_key] = [full_key]
+                    else:
+                        custom_frequency_attrs[freq_key].append(full_key)
+                        if custom_frequency_counts[freq_key] != attr_count:
+                            raise ValueError(
+                                f"Custom attributes with frequency '{freq_key}' have inconsistent counts: "
+                                f"expected {custom_frequency_counts[freq_key]} (from {custom_frequency_attrs[freq_key][0]}), "
+                                f"but '{full_key}' has {attr_count} values."
+                            )
```

Ruff (0.14.10)

Avoid specifying long messages outside the exception class

(TRY003)